IDCOM-2102 Controller support

dankelleher opened the pull request:
Cryptid v2 now finally supports controllers. Once we have client docs I will add this there, but example usage is:

Given did:sol:alice, a controller of did:sol:bob:

```
const cryptid = await Cryptid.buildFromDID(
        "did:sol:bob",
        aliceWallet,
        {  connection: provider.connection }
      ).controlWith("did:sol:alice");
```
This also works with controller chaining, so if did:sol:carol is a controller of did:sol:alice, you can do:

```
const cryptid = await Cryptid.buildFromDID(
        "did:sol:bob",
        carolWallet,
        {  connection: provider.connection }
      )
   .controlWith("did:sol:alice")
   .controlWith("did:sol:carol")
```

It is also possible to create a cryptid account for a DID as a controller:
```
Builder.createFromDID(controlledDid, controllerAuthority, { ..., controllerChain: [controllerDid]})
```
This API is different to the .controlWith() API for building cryptid clients, and therefore could do with being rethought a little perhaps.

dankelleher committed:
IDCOM-2102 WIP Controller support

dankelleher committed:
Merge branch 'anchor_rewrite' into feature/IDCOM-2102__Controller_support

dankelleher committed:
IDCOM-2102 Lint fixes

dankelleher committed:
IDCOM-2102 Update to support generative DIDs in the controller chain.

JulianSpring commented on programs/cryptid/src/instructions/create.rs:
```diff
@@ -46,7 +49,13 @@ pub fn create(
     // note - cryptid does not need to check that the chain is valid, or even that they are DIDs
     // sol_did does that
-    let controlling_did_accounts = ctx.remaining_accounts.iter().collect::<Vec<&AccountInfo>>();
+    // let controlling_did_accounts = ctx.remaining_accounts.iter().collect::<Vec<&AccountInfo>>();
```
Can we remove this?

dankelleher replied:
Done - thanks

JulianSpring commented on programs/cryptid/src/instructions/direct_execute.rs:
```diff
@@ -83,7 +81,19 @@ pub fn direct_execute<'a, 'b, 'c, 'info>(
+            )
+        })
+        .collect::<Vec<(&AccountInfo, Pubkey)>>();
+    // .as_slice();
```
Can we remove this?

dankelleher replied:
Done - thanks

dankelleher committed:
IDCOM-2102 Fixes to support chained controllers

dankelleher committed:
IDCOM-2102 Review Fix

rado0x54 commented on packages/client/core/src/api/simpleCryptidClient.ts:
```diff
@@ -17,7 +17,7 @@ export class SimpleCryptidClient extends AbstractCryptidClient {
 
-  as(controlledDid: string): ControlledCryptidClient {
-    return new ControlledCryptidClient(controlledDid, this, this.options);
+  controlWith(controllerDid: string): CryptidClient {
```
actually found the `as` quite elegant, but understand that this is more descriptive.

rado0x54 commented on packages/client/core/src/lib/CryptidTransaction.ts:
```diff
@@ -117,15 +136,75 @@ export class CryptidTransaction {
     );
   }
 
   // TODO move transactionAccountAddress into constructor?
```
This TODO was separated from its previous function.

rado0x54 commented:
As mentioned in the meeting, there is come TODOs left around propose execute tests still having `Buffer.from([]), // TODO, support controller chain,`

dankelleher committed:
IDCOM-2102 Add support for creating cryptid accounts via a controller. Right now the API for this operation looks different to the other cases. instead of cryptid.controlledBy(...), instead we are doing:

Builder.createFromDID(controlledDid, controllerAuthority, { ..., controllerChain: [controllerDid]})

This is a little messy and could do with being rethought a little perhaps
